libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -0,0 +1,177 @@
+            this._setUserLanguage = setUserLanguage ?? throw new ArgumentNullException(nameof(setUserLanguage)); 
+        }
+
+        /// <summary>
+        /// Incoming activity
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="next"></param>
+        /// <returns></returns>       
+        public async Task OnProcessRequest(ITurnContext context, MiddlewareSet.NextDelegate next)
+        {
+            IMessageActivity message = context.Activity.AsMessageActivity();
+            if (message != null)
+            {
+                if (!String.IsNullOrWhiteSpace(message.Text))
+                {
+
+                    var languageChanged = false;
+
+                    if (_setUserLanguage != null)
+                    {
+                        languageChanged = await _setUserLanguage(context);
+                    }
+
+                    if (!languageChanged)
+                    {
+                        // determine the language we are using for this conversation
+                        var sourceLanguage = "";
+                        if (_getUserLanguage == null)
+                            sourceLanguage = await Task.Run(() => _translator.Detect(message.Text)); // context.Conversation.Data["Language"]?.ToString() ?? this._nativeLanguages.FirstOrDefault() ?? "en";
+                        else
+                        {
+                            sourceLanguage = _getUserLanguage(context);
+                        } 
+                        if (_patterns.ContainsKey(sourceLanguage) && _patterns[sourceLanguage].Count>0)
+                        {
+                            this._translator.SetPostProcessorTemplate(_patterns[sourceLanguage]);
+                        }
+                        if (!_nativeLanguages.Contains(sourceLanguage))
+                        {
+                            var translationContext = new TranslationContext
+                            {
+                                SourceText = message.Text,
+                                SourceLanguage = sourceLanguage,
+                                TargetLanguage = (this._nativeLanguages.Contains(sourceLanguage)) ? sourceLanguage : this._nativeLanguages.FirstOrDefault() ?? "en"
+                            };
+
+                            // translate to bots language
+                            if (translationContext.SourceLanguage != translationContext.TargetLanguage)
+                                await TranslateMessageAsync(context, message, translationContext.SourceLanguage, translationContext.TargetLanguage).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+            await next().ConfigureAwait(false);
+        }
+
+
+        private static readonly Regex UrlRegex = new Regex(@"(https?://[^\s]*)", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Multiline);
```

Comment:
BUGS. **_Please don't do this_**. Use the System.URI class. 

There are soooooooo many possible bugs doing this.

Reply:
I will remove URLs and mentions checks as it is handled internally in Microsoft Translator API